src/pcap/api.go

```diff
@@ -0,0 +1,650 @@
+	for name, handler := range api.handlers {
+		if handler.canHandle(capture) {
+			log.Sugar().Debugf("Resolving agent endpoints via handler %s for capture %s", name, capture)
+
+			agents, err := handler.handle(capture, log)
+			if err != nil {
+				return nil, fmt.Errorf("error while handling %v via %s: %w", capture, name, err)
+			}
+
+			return agents, nil
+		}
+	}
+
+	return nil, fmt.Errorf("no handler for %v", capture)
+}
+
+func checkAgentStatus(statusRes *StatusResponse, err error, target AgentEndpoint) error {
+	if err != nil {
+		err = fmt.Errorf("status request finished with error for '%s': %w", target, err)
+		return err
+	}
+
+	if !(statusRes.Healthy) {
+		statusErr := fmt.Errorf("agent unhealthy '%s': %s", target, statusRes.Message)
+		return statusErr
+	}
+
+	if CompatibilityLevel > statusRes.CompatibilityLevel {
+		statusErr := fmt.Errorf("incompatible versions for '%s': expected compatibility level %d+ but got %d ", target, CompatibilityLevel, statusRes.CompatibilityLevel)
+		return statusErr
+	}
+	return nil
+}
+
+// Takes the data received for each of the pcap-agents and merges it into the resulting channel
+// The resulting channel is unbuffered.
+// inspired by: https://go.dev/blog/pipelines
+func mergeResponseChannels(cs []<-chan *CaptureResponse, bufSize int) <-chan *CaptureResponse {
+	var wg sync.WaitGroup
+	out := make(chan *CaptureResponse, bufSize)
+
+	// Start an output goroutine for each input channel in cs.  output
+	// copies values from c to out until c is closed, then calls wg.Done.
+	output := func(c <-chan *CaptureResponse) {
+		for n := range c {
+			out <- n
+		}
+		wg.Done()
+	}
+	wg.Add(len(cs))
+	for _, c := range cs {
+		go output(c)
+	}
+
+	// Start a goroutine to close out once all the output goroutines are done.
+	go func() {
+		wg.Wait()
+		close(out)
+	}()
+	return out
+}
+
+type streamPrep struct {
+}
+
+// prepareStreamToTarget creates a client connection to the given target, contacts the client API for the Agent service
+// to start the Capture.
+func (p *streamPrep) prepareStreamToTarget(ctx context.Context, req *CaptureOptions, target AgentEndpoint, creds credentials.TransportCredentials, log *zap.Logger) (captureReceiver, error) {
+	cc, err := grpc.Dial(target.String(), grpc.WithTransportCredentials(creds))
+	if err != nil {
+		err = fmt.Errorf("start capture from '%s': %w", target, err)
+		return nil, err
+	}
+
+	agent := NewAgentClient(cc)
+
+	statusRes, err := agent.Status(ctx, &StatusRequest{})
+	err = checkAgentStatus(statusRes, err, target)
+	if err != nil {
+		return nil, err
+	}
+
+	agentContext := context.Background()
+	vcapID, err := vcapIDFromOutgoingCtx(ctx)
+	if err != nil {
+		// TODO impelement error handling
+	}
+	agentContext, log = setVcapID(agentContext, log, vcapID)
+	captureStream, err := agent.Capture(agentContext)
+
+	if err != nil {
+		convertStatusCodeToMsg(err, target)
+		return nil, err
+	}
+
+	patchedFilter, err := patchFilter(req.Filter)
```

Comment:
I don't see the need for a new variable.

```suggestion
	req.Filter, err := patchFilter(req.Filter)
```

Reply:
I think it is ok to have a new variable and in error case not change the req.Filter.